File: classic150/Solution0199.cpp
#include <vector>
#include <queue>
#include <stack>
using namespace std;
// 给定一个二叉树的 根节点 root，想象自己站在它的右侧，按照从顶部到底部的顺序，返回从右侧所能看到的节点值。
class Solution
{
private:
    struct TreeNode
    {
        int val;
        TreeNode *left;
        TreeNode *right;
        TreeNode() : val(0), left(nullptr), right(nullptr) {}
        TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
        TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
    };

public:
    // 使用层次遍历，但是压入节点的时候先压入右节点，这样可以保证先遍历到右侧节点
    vector<int> rightSideView(TreeNode *root)
    {
        vector<int> result;
        if (!root)
            return result;
        queue<TreeNode *> q;
        q.push(root);
        while (!q.empty())
        {
            int size = q.size();
            for (int i = 0; i < size; i++)
            {
                TreeNode *node = q.front();
                q.pop();
                if (i == 0) // 第一个访问就是右侧可见节点
                    result.push_back(node->val);
                if (node->right)
                    q.push(node->right);
                if (node->left)
                    q.push(node->left);
            }
        }
        return result;
    }

    // 直接存储每个节点所在层
    vector<int> rightSideView2(TreeNode *root)
    {
        vector<int> result;
        if (!root)
            return result;
        queue<pair<TreeNode *, int>> q; // 存储节点和所在层级
        q.push({root, 0});
        while (!q.empty())
        {
            int size = q.size();
            while (size--)
            {
                auto [node, level] = q.front();
                q.pop();
                if (level == result.size())
                    result.push_back(node->val);
                if (node->right)
                    q.push({node->right, level + 1});
                if (node->left)
                    q.push({node->left, level + 1});
            }
        }
        return result;
    }
};

class Solution2
{
private:
    struct TreeNode
    {
        int val;
        TreeNode *left;
        TreeNode *right;
        TreeNode() : val(0), left(nullptr), right(nullptr) {}
        TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
        TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
    };

public:
    // 使用深度优先遍历
    vector<int> rightSideView(TreeNode *root)
    {
        vector<int> result;
        dfs(root, 0, result);
        return result;
    }

private:
    void dfs(TreeNode *node, int level, vector<int> &result)
    {
        if (!node)
            return;
        if (level == result.size()) // 说明该层是第一次访问
            result.push_back(node->val);
        dfs(node->right, level + 1, result);
        dfs(node->left, level + 1, result);
    }

public:
    vector<int> rightSideView(TreeNode *root)
    {
        vector<int> result;
        stack<pair<TreeNode *, int>> stk; // 栈存储节点和其所在层级
        stk.push({root, 0});
        while (!stk.empty())
        {
            auto [node, level] = stk.top();
            stk.pop();
            if (level == result.size())
                result.push_back(node->val);
            if (node->right)
                stk.push({node->right, level + 1});
            if (node->left)
                stk.push({node->left, level + 1});
        }
        return result;
    }
};

File: anew/Solution0199.java
package anew;

import java.util.List;
import java.util.ArrayList;
import java.util.Queue;
import java.util.Stack;
import java.util.LinkedList;

// 给定一个二叉树的 根节点 root，想象自己站在它的右侧，按照从顶部到底部的顺序，返回从右侧所能看到的节点值。

public class Solution0199 {
    private static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode(int val) {
            this.val = val;
        }
    }

    public List<Integer> rightSideView(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        if (root == null) {
            return result;
        }
        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);
        while (!queue.isEmpty()) {
            int levelSize = queue.size();
            for (int i = 0; i < levelSize; i++) {
                TreeNode currentNode = queue.poll();
                if (i == levelSize - 1) { // 每层的最后一个节点就是右侧视图中的节点
                    result.add(currentNode.val);
                }
                if (currentNode.left != null) {
                    queue.offer(currentNode.left);
                }
                if (currentNode.right != null) {
                    queue.offer(currentNode.right);
                }
            }
        }
        return result;
    }

    private record NodeWithLevel(TreeNode node, int level) {
    }

    public List<Integer> rightSideView2(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        if (root == null) {
            return result;
        }
        Queue<NodeWithLevel> queue = new LinkedList<>();
        queue.offer(new NodeWithLevel(root, 0));
        while (!queue.isEmpty()) {
            NodeWithLevel current = queue.poll();
            TreeNode currentNode = current.node();
            int currentLevel = current.level();
            if (currentLevel == result.size()) { // 每层的第一个节点就是右侧视图中的节点
                result.add(currentNode.val);
            }
            if (currentNode.right != null) {
                queue.offer(new NodeWithLevel(currentNode.right, currentLevel + 1));
            }
            if (currentNode.left != null) {
                queue.offer(new NodeWithLevel(currentNode.left, currentLevel + 1));
            }
        }
        return result;
    }

    public List<Integer> rightSideView3(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        if (root == null) {
            return result;
        }
        Stack<NodeWithLevel> stack = new Stack<>();
        stack.push(new NodeWithLevel(root, 0));
        while (!stack.isEmpty()) {
            NodeWithLevel current = stack.pop();
            TreeNode currentNode = current.node();
            int currentLevel = current.level();
            if (currentLevel == result.size()) { // 每层的第一个节点就是右侧视图中的节点
                result.add(currentNode.val);
            }
            if (currentNode.left != null) {
                stack.push(new NodeWithLevel(currentNode.left, currentLevel + 1));
            }
            if (currentNode.right != null) {
                stack.push(new NodeWithLevel(currentNode.right, currentLevel + 1));
            }
        }

        return result;
    }

    public List<Integer> rightSideView4(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        dfs(root, 0, result);
        return result;
    }

    private void dfs(TreeNode node, int level, List<Integer> result) {
        if (node == null) {
            return;
        }
        if (level == result.size()) { // 每层的第一个节点就是右侧视图中的节点
            result.add(node.val);
        }
        // 先右后左遍历，这样每层的第一个节点就是右侧视图中的节点
        dfs(node.right, level + 1, result);
        dfs(node.left, level + 1, result);
    }

    public static void main(String[] args) {
        Solution0199 solution = new Solution0199();
        TreeNode root = new TreeNode(1);
        root.left = new TreeNode(2);
        root.right = new TreeNode(3);
        root.left.right = new TreeNode(5);
        root.right.right = new TreeNode(4);
        System.out.println(solution.rightSideView(root)); // 输出: [1, 3, 4]
        System.out.println(solution.rightSideView2(root)); // 输出: [1, 3, 4]
        System.out.println(solution.rightSideView3(root)); // 输出: [1, 3, 4]
        System.out.println(solution.rightSideView4(root)); // 输出: [1, 3, 4]
    }

}
